android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+        }
+
+        if (notificationCache.prevNotificationID != -1) {
+            NotificationManagerCompat.from(context).cancel(notificationCache.prevNotificationID);
+        }
+
+        builder.setStyle(messagingStyle);
+        notificationCache.prevNotificationID = notificationID;
+    }
+
+    /**
+     * Check if we are showing a notification related to a reportID.
+     * If not, create a new NotificationCache so we can build a conversation notification
+     * as the messages come.
+     *
+     * @param reportID Report ID.
+     * @return Notification Cache.
+     */
+    private NotificationCache findOrCreateNotificationCache(int reportID) {
+        NotificationCache notificationCache = cache.get(reportID);
+
+        if (notificationCache == null) {
+            notificationCache = new NotificationCache();
+            cache.put(reportID, notificationCache);
+        }
+
+        return notificationCache;
+    }
+
+    /**
+     * Remove the notification data from the cache when the user dismisses the notification.
+     *
+     * @param message Push notification's message
+     */
+    public void onDismissNotification(PushMessage message) {
+        try {
+            JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
+            int reportID = payload.get("reportID").getInt(-1);
+
+            if (reportID == -1) {
+                return;
+            }
+
+            cache.remove(reportID);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to delete conversation cache");
+        }
+    }
+
+    @NonNull
+    private IconCompat fetchIcon(@NonNull String urlString, @DrawableRes int fallbackId) {
+        // TODO: Add disk LRU cache
+
+        URL parsedUrl = null;
+        try {
+            parsedUrl = urlString == null ? null : new URL(urlString);
+        } catch (MalformedURLException e) {
+            Log.e(TAG, "Failed to resolve URL " + urlString, e);
+        }
+
+        if (parsedUrl != null) {
+            WindowManager window = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+            DisplayMetrics dm = new DisplayMetrics();
+            window.getDefaultDisplay().getMetrics(dm);
+
+            final int reqWidth = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, MAX_ICON_SIZE_DPS, dm);
+            final int reqHeight = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, MAX_ICON_SIZE_DPS, dm);
+
+            final URL url = parsedUrl;
+            Future<Bitmap> future = executorService.submit(() -> ImageUtils.fetchScaledBitmap(context, url, reqWidth, reqHeight));
+
+            try {
+                Bitmap bitmap = future.get(MAX_ICON_FETCH_WAIT_TIME_SECONDS, TimeUnit.SECONDS);
+                return IconCompat.createWithBitmap(bitmap);
+            } catch (InterruptedException e) {
+                Log.e(TAG,"Failed to fetch icon", e);
+                Thread.currentThread().interrupt();
```

Comment:
What is this for?

Reply:
No idea. I copied this function from a Gist that Urban Airship provided us.